app/services/attachments/concatenation_service.rb

```diff
@@ -0,0 +1,153 @@
+
+      if attachments.length != attachment_ids.length
+        raise AttachmentConcatenationError,
+              I18n.t('services.attachments.concatenation.incorrect_attachable')
+      end
+
+      if attachments.length.positive?
+        is_paired_end = attachments.first.metadata.key?('type')
+        is_single_end = !attachments.first.metadata.key?('type')
+
+        validate_same_types(attachments, is_paired_end, is_single_end)
+
+        if is_paired_end
+          forward_reads = []
+          reverse_reads = []
+          attachments.each do |attachment|
+            if attachment.metadata['direction'] == 'forward'
+              forward_reads << attachment
+            elsif attachment.metadata['direction'] == 'reverse'
+              reverse_reads << attachment
+            end
+          end
+          concatenate_paired_end_reads(forward_reads, reverse_reads)
+        elsif is_single_end
+          concatenate_single_end_reads(attachments)
+        end
+
+        # if option is selected then destroy the original files
+        attachments.each(&:destroy!) if concatenation_params[:delete_originals]
+      end
+    rescue Attachments::ConcatenationService::AttachmentConcatenationError => e
+      attachable.errors.add(:base, e.message)
+      attachable
+    end
+
+    private
+
+    # Validates attachments to ensure they are of the correct type.
+    # Files can only be concatenated as follows. No mixing of types:
+    # Paired-end -> Paired-end
+    # Single-end -> Single-end
+    def validate_same_types(attachments, is_paired_end, is_single_end) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+      valid_type = true
+
+      attachments.each do |attachment|
+        if is_single_end
+          valid_type = !attachment.metadata.key?('type')
+        elsif is_paired_end
+          valid_type = attachment.metadata.key?('type')
+        end
+      end
+
+      if valid_type
+        expected_extension = attachments.first.file.filename.to_s.partition('.').last
+        attachments.each do |attachment|
+          extension = attachment.file.filename.to_s.partition('.').last
+          next unless extension != expected_extension
+
+          valid_type = false
+          raise AttachmentConcatenationError,
+                I18n.t('services.attachments.concatenation.incorrect_fastq_file_types')
+        end
+      end
+
+      return if valid_type
+
+      raise AttachmentConcatenationError,
+            I18n.t('services.attachments.concatenation.incorrect_file_types')
+    end
+
+    # Concatenates the single end reads into a single-end file
+    def concatenate_single_end_reads(attachments)
+      basename = concatenation_params[:basename] || 'concatenated_file'
+      extension = attachments.first.file.filename.to_s.partition('.').last
+
+      extension = 'fastq.gz' if extension == 'gz'
```

Comment:
**question (non-blocking):** Do you think that it would help if we added `compression` metadata to `attachment` and have it `none` if no compression and `gzip` if `.gz`?

Reply:
Yup definitely would. Will be doing this in a separate commit

Reply:
Added compression metadata to attachments in fc5a524e4a03e80f0bd3b6e0021f4cc842a0acac